pkg/search/postgres/storecache.go

```diff
@@ -0,0 +1,403 @@
+func (c *cachedStore[T, PT]) DeleteByQuery(ctx context.Context, query *v1.Query) error {
+	objs, dbFetchErr := c.underlyingStore.GetByQuery(ctx, query)
+	if dbFetchErr != nil {
+		return dbFetchErr
+	}
+	identifiersToRemove := make([]string, 0, len(objs))
+	for _, obj := range objs {
+		identifiersToRemove = append(identifiersToRemove, c.pkGetter(obj))
+	}
+	dbRemoveErr := c.underlyingStore.DeleteMany(ctx, identifiersToRemove)
+	if dbRemoveErr != nil {
+		return dbRemoveErr
+	}
+	defer c.setCacheOperationDurationTime(time.Now(), ops.Remove)
+	c.cacheLock.Lock()
+	defer c.cacheLock.Unlock()
+	for _, id := range identifiersToRemove {
+		delete(c.cache, id)
+	}
+	return nil
+}
+
+// GetIDs returns all the IDs for the store.
+func (c *cachedStore[T, PT]) GetIDs(ctx context.Context) ([]string, error) {
+	defer c.setCacheOperationDurationTime(time.Now(), ops.GetAll)
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	result := make([]string, 0, len(c.cache))
+	walkErr := c.walkCacheNoLock(ctx, func(obj PT) error {
+		result = append(result, c.pkGetter(obj))
+		return nil
+	})
+	if walkErr != nil {
+		return nil, walkErr
+	}
+	return result, nil
+}
+
+// GetAll retrieves all objects from the store.
+//
+// Deprecated: This can be dangerous on high cardinality stores consider Walk instead.
+func (c *cachedStore[T, PT]) GetAll(ctx context.Context) ([]PT, error) {
+	defer c.setCacheOperationDurationTime(time.Now(), ops.GetAll)
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	result := make([]PT, 0, len(c.cache))
+	walkErr := c.walkCacheNoLock(ctx, func(obj PT) error {
+		result = append(result, obj)
+		return nil
+	})
+	if walkErr != nil {
+		return nil, walkErr
+	}
+	return result, nil
+}
+
+func (c *cachedStore[T, PT]) walkCacheNoLock(ctx context.Context, fn func(obj PT) error) error {
+	for _, obj := range c.cache {
+		if !c.isReadAllowed(ctx, obj) {
+			continue
+		}
+		fnErr := fn(obj)
+		if fnErr != nil {
+			return fnErr
+		}
+	}
+	return nil
+}
+
+func (c *cachedStore[T, PT]) isReadAllowed(ctx context.Context, obj PT) bool {
+	return c.isActionAllowed(ctx, storage.Access_READ_ACCESS, obj)
+}
+
+func (c *cachedStore[T, PT]) isWriteAllowed(ctx context.Context, obj PT) bool {
+	return c.isActionAllowed(ctx, storage.Access_READ_WRITE_ACCESS, obj)
+}
+
+func (c *cachedStore[T, PT]) isActionAllowed(ctx context.Context, action storage.Access, obj PT) bool {
+	if c.hasPermissionsChecker() {
+		var allowed bool
+		var err error
+		switch action {
+		case storage.Access_READ_ACCESS:
+			allowed, err = c.permissionChecker.ReadAllowed(ctx)
+		case storage.Access_READ_WRITE_ACCESS:
+			allowed, err = c.permissionChecker.WriteAllowed(ctx)
+		default:
+			return false
+		}
+		if err != nil {
+			return false
+		}
+		return allowed
+	}
+	scopeChecker := sac.GlobalAccessScopeChecker(ctx)
+	scopeChecker = scopeChecker.AccessMode(action)
+	scopeChecker = scopeChecker.Resource(c.targetResource)
```

Comment:
This could be simplified to the below code, right?

```suggestion
	scopeChecker := sac.GlobalAccessScopeChecker(ctx).AccessMode(action).Resource(c.targetResource)
```